ui/button.cpp

```diff
@@ -0,0 +1,83 @@
+#include "button.h"
+
+#include "pong.h"
+#include "uielement.h"
+
+namespace pong
+{
+
+Button::Button(float width, float height) :
+    mRectangle(width, height)
+{
+    mColliderBox.Resize(width, height);
+}
+
+void Button::Render() const
+{
+    mRectangle.Draw(mPosition);
+}
+
+UIElementType Button::GetType() const
+{
+    return UIElementType::Button;
+}
+
+void Button::AddListener(ButtonEvent event, std::function<void()> callback)
+{
+    ASSERT(event < ButtonEvent::EVENTS_COUNT);
```

Comment:
is this even possible? As long as someone isn't getting crazy and doing the cast to and from integers, this would be hard. `enum class`, which you've been good about using is much more resilient to issues like this than c style enums.

Reply:
Well I guess this would just be preventing from doing something like AddListener(ButtonEvent::EVENTS_COUNT, []() {});
directly passed into the function.